### ISSUE TITLE
Audio

### ISSUE THREAD

**Issue by john-peterson** [mirror/dolphin-emu#37]
### Problem

The call stack of the preceding code with symbols is removed if a crash occur in OpenAL32.dll or wrap_oal.dll

```
Unflagged   >   23352   0   Worker Thread   Audio - OpenAL      Normal
wrap_oal.dll    C:\document\source\dolphin\repo\Binary\x64\wrap_oal.dll N/A N/A Cannot find or open the PDB file        95  2.01.8.1    2007-07-09 19:47    00000000275D0000-0000000027815000*  [20196] DolphinDF.exe: Native   
message in icon column  "the module did not load at the default load address"
No disassembly available.
Call Stack      empty
```

The message "the module did not load at the default load address" is present when it operates as intended too

The crash occur
- before OpenALStream::SoundLoop execution (determined because the thread doesn't pause on breakpoints)
- around 1 s after these messages

```
19:51:482 Src\OpenALStream.cpp:26 W[Audio]: Found device
19:51:483 Src\OpenALStream.cpp:28 W[Audio]:     OpenAL Soft
19:51:483 Src\OpenALStream.cpp:28 W[Audio]:     Generic Software
19:51:483 Src\OpenALStream.cpp:32 N[Audio]: Select device Generic Software
19:51:545 Src\Thread.cpp:118 N[COMMON]: Created "Audio - OpenAL" 8168
```

The problem also occur with
- open OpenAL, dynamic build
- open OpenAL, static build
- non-open OpenAL, dynamic build
### Solution

If a crash on function entry is caused by a stack size overflow the solution is
- move arrays from stack to heap

The call stack isn't removed if a crash occur in a module without symbols
### Problem

The problem in the topic "Empty call stack" occur on consecutive runs
- after the OpenALStream object has been removed before OpenALStream::SoundLoop has exited, because of a [deadlock](https://github.com/mirror/dolphin-emu/issues/35) that's a result of an endless loop that's a result of an inadequate buffer size
- when run from VS
- not when VS is attached after the process is started
- not when VS isn't attached
- not when run from VS when using the 32 instead of 64 bit build
- regardless of
  - build configuration
  - the setting Fastmem, DSPThread

for around 50 runs, of wich around 10 are consecutive with the problem

i.o.w. an environment state related to f.e. sound device allocation that it's sensitive to has changed

this state is without apparent reason undone so that the crash doesn't occur

The state that cause the problem isn't easily located because
- the problem occur before the first breakpoint in OpenALStream::SoundLoop
- the call stack is empty as described in the topic "Empty call stack"
- it occur with the open OpenAL too

The most likely cause is a OpenAL build setting that affect debugging. However, OpenAl build settings related to debugging information have been altered without solving the problem
### Solution

The solution isn't clear because it has occured with these OpenAL combinations
- open OpenAL, dynamic build
- open OpenAL, static build
- non-open OpenAL, dynamic build
### Problem

At one observation, breaking at

```
memset(uiBuffers, 0, numBuffers * sizeof(ALuint));
```

and pressing step (F11) open the dialog "Find source: memset.asm"

```
f:\dd\vctools\crt_bld\SELF_64_amd64\crt\src\amd64\memset.asm
```

At two observations (in DebugFast and Debug), step (F11)
- step past 200 lines from the the top to the lower code
  
  ```
  for (u32 i = 0; i < numSamples * STEREO_CHANNELS; ++i)
      dest[i] = (float)realtimeBuffer[i] / (1 << 16);
  
  for (u32 i = 0; i < nSamples * STEREO_CHANNELS; ++i)
      stereo[i] = (short)((float)sampleBuffer[i] * (1 << 16));
  ```
## TracePoint problem

A TracePoint (of f.e. `{i} {float32_capable} {surround_capable}`) in `OpenALStream::SoundLoop` return messages similar to

```
Function: OpenALStream::SoundLoop(void), Thread: 0x5D50 Audio - OpenAL CXX0017: Error: symbol "i" not found
Function: OpenALStream::SoundLoop(void), Thread: 0x5E0 Audio - OpenAL CXX0017: Error: symbol "float32_capable" not found CXX0017: Error: symbol "surround_capable" not found 
```

This occur in
- VS 10
- VS 11

**Comment by john-peterson** [comment]
## Build OpenAL32.dll

This build OpenAL32.dll 

```
git clone git://github.com/mirror/openal-soft.git
mkdir 64 && cd 64
cmake .. -G "Visual Studio 10 Win64"
call "%VS100COMNTOOLS%..\..\VC\vcvarsall.bat"
msbuild OpenAL.sln /t:ALL_BUILD /p:Configuration=Debug;Platform=x64
msbuild OpenAL.sln /t:ALL_BUILD /p:Configuration=Release;Platform=x64
# 32
mkdir 32 && cd 32
cmake .. -G "Visual Studio 10"
msbuild OpenAL.sln /t:ALL_BUILD /p:Configuration=Debug;Platform=Win32
msbuild OpenAL.sln /t:ALL_BUILD /p:Configuration=Release;Platform=Win32
# static
cmake .. -G "Visual Studio 10 Win64" -D LIBTYPE=STATIC
cmake .. -G "Visual Studio 10" -D LIBTYPE=STATIC
```

It doesn't create a file called soft_oal.dll because the function of soft_oal.dll is in it

**Comment by john-peterson** [comment]
## Changing from a non-open to open OpenAL

### Patch

https://github.com/john-peterson/dolphin-emu/compare/audio

### Problem

These files are in Externals/OpenAL/Win64 (since 6df1dac)
- OpenAL32.dll: non-open code from OpenAL11CoreSDK.zip
- soft_oal.dll
- wrap_oal.dll

### Solution

Remove
- wrap_oal.dll becase it's non-open
- EFX-Util.lib because it's unused

rename
- soft_oal.dll to OpenAL32.dll i.o.w. the name the repo build use

because
- if the source is known a problem can be located
- the "OpenAL Soft" device is adequate because the OpenAL backend doesn't benefit from the device "Generic Hardware" or another device

The file
- OpenAL32.pdb

isn't added because
- it's large and it increase the main repo size because the dependencies [aren't in submodules](https://github.com/mirror/dolphin-emu/issues/7)
- if there's a problem it can easily be added

### Discussion

If soft_oal.dll and wrap_oal.dll are mutually exclusive the previous should be used because its source is known

> [08:32] @skid_au yep, that's fine. sounds good

**Comment by john-peterson** [comment]
## Clarifying the OpenAL loop

### Patch

https://github.com/john-peterson/dolphin-emu/commit/a2ec263d67864ddf4c5fcaf4fe15067c8e92091d

### Problem

#### Names for numbers

Some numbers that should have a name doesn't have that

This is a problem because
- that makes it harder to understand the code

#### Inadequate buffer size

The buffer

```
short stereo[OAL_MAX_SAMPLES * 2 * 2 * OAL_MAX_BUFFERS];
```

was before this patch half of the max size resulting in the message

```
Run-Time Check Failure #2 - Stack around the variable 'stereo' was corrupted.
```

that's noticed when
- stopping Wii Sports in a Debug build

because
- that result in a `soundTouch.receiveSamples` size that's larger than half its maximum

### Solution

#### `sampleBuffer` size

Removing `SIZE_FLOAT` multiplication from this `sampleBuffer` code

```
soundtouch::SAMPLETYPE sampleBuffer[OAL_MAX_SAMPLES * SIZE_FLOAT * SURROUND_CHANNELS * OAL_MAX_BUFFERS];
nSamples = soundTouch.receiveSamples(sampleBuffer, OAL_MAX_SAMPLES * SIZE_FLOAT * OAL_MAX_BUFFERS);
```

because
- the number of samples is the number of 2 channel 4 byte/channel (float) frames (rather than number of bytes)
- there're no logical reason for this multiplication

Changing the code to

```
unsigned int nSamples = soundTouch.receiveSamples(sampleBuffer, OAL_MAX_SAMPLES * numBuffers);
```

rather than

```
unsigned int nSamples = soundTouch.receiveSamples(sampleBuffer, OAL_MAX_SAMPLES * OAL_MAX_BUFFERS);
```

because

it shouldn't be possible to recieve more buffers than the number played in

```
if (iBuffersFilled == numBuffers)
    alSourcePlay(uiSource);
```

#### Names for numbers

Adding names to numbers

```
#define STEREO_CHANNELS         2
#define SIZE_SHORT              2
#define FRAME_STEREO_SHORT      STEREO_CHANNELS * SIZE_SHORT
#define FRAME_STEREO_FLOAT      STEREO_CHANNELS * SIZE_FLOAT
#define FRAME_SURROUND_FLOAT    SURROUND_CHANNELS * SIZE_FLOAT
```

because
- that makes it easier to understand what the number mean

#### Refactoring

Moving

```
surround_capable
float32_capable
```

to top of function because
- that's better organisation

Replacing a bracket with

```
if (nSamples <= minSamples)
    continue;
```

because
- that reduce indentation

### Test

This is tested with
- [files](https://www.dropbox.com/sh/2z5p162ovqtbn17/y7EfCJd1Oi/file/dolphin/test)

in
- SSBM because it play continuous sound directly after start without user input

with

the `User/Config/Dolphin.ini` settings

```
DSPHLE = False

DPL2Decoder = True
DPL2Decoder = False

Latency = 30
Latency = 0
```

CPU

```
sudo dmidecode --type processor | grep "Version"
   Version: Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz
```

### Merge status

> [02:33] <skidau> JPeterson, your openal changes sound fine. merge those in. keep in mind what KC said about openal soft
> 
> [02:56] <skidau> said that replacing OpenAL32.dll might break hw accel for some

**Comment by john-peterson** [comment]
## Adding OpenAL source to project

### Patch

https://github.com/john-peterson/dolphin-emu/compare/audio

### Problem

The solution for the problem in the topic "Missing call stack" isn't clear

### Solution

Adding the OpenAL source to the project because that makes it easier to
- debug code that use it
- change build settings that affect debugging

**Comment by john-peterson** [comment]
## soft_oal.dll

soft_oal.dll is a driver that Creative's OpenAL32.dll router will use if it's in the same folder

Creative's OpenAL32.dll ALDeviceList::GetDefaultDevice use this priority
- Generic Software (wrap_oal.dll)
- OpenAL Soft (soft_oal.dll)

### Discussion

How is soft_oal.dll built?

It's created by building the repo and renaming the resulting OpenAL32.dll to soft_oal.dll

What's the purpose of using soft_oal.dll through Creative's OpenAL32.dll instead of the repo build of OpenAL32.dll?

Why does openal-soft-1.15.1-bin.zip contain a file caleld `soft_oal.dll` instead of `OpenAL32.dll`? that's the name created by the repo build

> [10:19] <+KittyCat> JPeterson, using soft_oal.dll with Creative's OpenAL32.dll allows other drivers to remain available (such as Generic Hardware, Generic Software, Creative's hardware drivers, Rapture3D, etc)

What's the benefit using the other drivers instead of OpenAL Soft?

> [10:24] <+KittyCat> JPeterson, Generic Hardware and Creative's hardware drivers can use hardware acceleration and hardware capabilities
> 
> [10:24] <+KittyCat> but there's not much benefit to Generic Software, and I don't really have an opinion to offer on Rapture3D since I've never tried it (it's not free)

## Select default OpenAL device

How's the default device that's selected by ALDeviceList::GetDefaultDevice ([here](https://github.com/mirror/dolphin-emu/blob/master/Source/Core/AudioCommon/Src/OpenALStream.cpp#L26)) selected?

> [11:04] <+KittyCat> JPeterson, looks like it's using normal openal enumeration
> [11:04] <+KittyCat> so the default will depend on the system and/or implementation

What does implementation refer to?

> [11:11] <+KittyCat> implementation basically means the driver
> [11:12] <+KittyCat> openal soft is an implementation. wrap_oal.dll is an implementation, as is 

Can I influence the default device selection through a setting in f.e. registry?

> [11:14] <+KittyCat> you can't influence which driver that the OpenAL32.dll router uses by default, although it should pay attention to whatever you have set as the default device for actual output
> [11:16] <+KittyCat> so it should use the default audio device as you have set in the control panel or wherever, but you can't tell it to use OpenAL Soft by default (it will try to use Generic Hardware or Generic Software first, if there isn't a real OpenAL hardware driver for it)

The devices in Playback in `mmsys.cpl` are
- Speakers
- AMD HDMI Output

The existing OpenAL devices are
- OpenAL Soft
- Generic Software

**Comment by john-peterson** [comment]
## Audio function description

This describe
- size type
- supported frame size

for the OpenAL loop functions becase it's benefical to know when changing is

#### SoundTouch



```
FIFOProcessor::putSamples
FIFOProcessor::receiveSamples
```
- 8 bytes
- 2 channel 4 byte/channel (float) frame

#### Dolby Pro Logic 2 decoder

```
dpl2decode
```
- in: 8 bytes, out: 8 \* 3 bytes
- in: 2 channel 4 byte/channel (float) frame, out: 6 channel 4 byte/channel (float) frame

#### Open Audio Library

```
alBufferData
```
- bytes
- 2 or 6 channel 4 or 2 byte/channel (float or signed integer) frame (AL_FORMAT_51CHN32, AL_FORMAT_STEREO_FLOAT32, AL_FORMAT_STEREO16)

**Comment by john-peterson** [comment]
## Listing found OpenAL devices

### Patch

https://github.com/john-peterson/dolphin-emu/commit/b0670cc4b8bc00fe30b48854cb8c722a6e426ee6

### Problem

The log display only the default device

```
08:19:451 Src\OpenALStream.cpp:28 W[Audio]: Found OpenAL device Generic Software
```

### Solution

List all devices because
- that allow investigating whether the appropriate device is selected

**Comment by john-peterson** [comment]
## Increasing the maximum OpenAL buffer size

### Commit

The commit has the same title as this post in https://github.com/john-peterson/dolphin-emu/compare/audio2

### Solution

#### Increasing the maximum buffer size

The maximum buffer size is increased from 32 to 512 because

`User/Config/Dolphin.ini` setting

```
Latency = 32
```

is sometimes not enough to prevent a buffer underrun as described in the topic "DTK audio stuttering"

> [04:51] @skid_au JPeterson, openal has got a maximum buffer limit of 32: http://stackoverflow.com/questions/2871905/openal-determine-maximum-sources

That Stack Overflow discussion is about sources rather than buffers

This code doesn't create an error when `numBuffers = 512`

```
alGenBuffers(numBuffers, (ALuint *)uiBuffers);
```

> [05:21] @skid_au it might one other operating systems or hardware or drivers tho?

Which OS, hardware or driver?

#### Equalising the buffer setting value with the internal value

The setting range is changed from "0 – (OAL_MAX_BUFFERS - 2)" to "2 - OAL_MAX_BUFFERS" because that
- remove the difference between the setting and the internal buffer size
- reduce amiguity about the the buffer size

#### Limiting setting range

```
numBuffers = Core::g_CoreStartupParameter.iLatency + 2; // OpenAL requires a minimum of two buffers
```

is changed to

```
numBuffers = Core::g_CoreStartupParameter.iLatency;
// OpenAL requires a minimum of two buffers
numBuffers = MathUtil::Trim(numBuffers, 2, OAL_MAX_BUFFERS);
```

because that disallow a buffer overflow as a result of an invalid `Latency` setting in `User/Config/Dolphin.ini`

#### Name for number

"30" in

```
wxSP_ARROW_KEYS, 0, 30);
```

is given a name

```
wxSP_ARROW_KEYS, 2, OAL_MAX_BUFFERS);
```

because
- that makes it clearer what the number mean

> [04:52] @skid_au also putting a openal specific setting in soundstream.h (which is generic) is not good

#### Moving array from stack to heap

The arrays

```
dpl2
stereo
```

are moved from the stack to the heap because
- they don't fit in the stack limit when OAL_MAX_BUFFERS is increased from 32 to 512

with the result that VS print this message when the stack allocation occur (when `OpenALStream::SoundLoop` is run)

> First-chance exception at 0x0000000001964E17 in Dolphin.exe: 0xC00000FD: Stack overflow (parameters: 0x0000000000000001, 0x0000000023A13000).

### Sound delay

A disadvantage with

`User/Config/Dolphin.ini` setting

```
Latency =256
```

is that it create a sound delay of around 2 s

because there are around 100 buffers/s

the buffers/s rate vary with the buffer size that vary with nSamples

**Comment by john-peterson** [comment]
## Dolby Pro Logic 2 decoder minimum sample size

### Test

The `User/Config/Dolphin.ini` setting

```
DPL2Decoder = True
Latency = 32
```

in

SSBM

create a delay (around 2 s) after `alSourcePlay` before audio is audible at the last of these log messages

```
static ALint iBuffersFilled_ = 0;
static ALint iBuffersProcessed_ = 0;
…
iBuffersProcessed_ += iBuffersProcessed;
…
iBuffersFilled_++;
NOTICE_LOG(CONSOLE, "iBuffersFilled: %d, %d | %d, %d", iBuffersFilled_, iBuffersFilled, iBuffersProcessed_, iBuffersProcessed);

40:45:556 Src\OpenALStream.cpp:340 N[CONSOLE]: alSourcePlay 2
40:45:592 Src\OpenALStream.cpp:328 N[CONSOLE]: alSourcePlay 1
40:50:378 Src\OpenALStream.cpp:324 N[CONSOLE]: iBuffersFilled: 174, 2 | 32475, 2
```

with

```
if (nSamples < minSamples)
    continue;
```

changed to

```
//if (nSamples < minSamples)
    //continue;
```

the equivalent log output has a cumulative iBuffersFilled around 10 times higher meaning that around 1/10 loops were skipped

```
42:35:101 Src\OpenALStream.cpp:340 N[CONSOLE]: alSourcePlay 2
42:38:884 Src\OpenALStream.cpp:324 N[CONSOLE]: iBuffersFilled: 1818, 18 | 35406, 18
```

### Problem

The code

```
if (nSamples < minSamples)
    continue;
```

 might be wrong because

`sampleBuffer` is always written at 0 by `FIFOSampleBuffer::receiveSamples` rather than written where the last write ended

```
uint FIFOSampleBuffer::receiveSamples(SAMPLETYPE *output, uint maxSamples)
    memcpy(output, ptrBegin(), channels * sizeof(SAMPLETYPE) * num);
```

but it shouldn't be progressed until `nSamples` accumulate to 240

### Solution

https://github.com/john-peterson/dolphin-emu/compare/audio5

Removing the sample minimum for `dpl2decode` because
- it's not clear that the `dpl2decode` output is different if less than `FWRDURATION` samples are used

### Alternative solution

https://github.com/john-peterson/dolphin-emu/compare/audio4

If the `FWRDURATION` sample minimum has a meaning

Placing the wait directly after

```
soundTouch.putSamples(dest, numSamples);
```

with the code

```
numSamples_ += numSamples;
if (numSamples_ < minSamples)
{
    continue;
}
numSamples_ = 0;
```

### Changing the `surround_capable` `minSamples` from 240 to 239

Because the sisze 240 should be included

**Comment by john-peterson** [comment]
## Fixing an unintended change in the OpenAL loop

### Patch

https://github.com/john-peterson/dolphin-emu/commit/c1fde1679479f09dc18cfa2e000cae53cc7656a2

### Merge status

> 09:23] @skid_au JPeterson, looks good. i'm happy with that

Do you want to commit the patch because I don't have [commit access](https://code.google.com/p/dolphin-emu/people/list?num=50&start=50)

> [09:43] @skid_au i'm not set up to commit that.. Sonicadvance1, you able to commit the patch?
> 
> [09:45] @Sonicadvance1 ah yea, give me a mom
> 
> [09:47] <irrawaddy> eb579e4 by Ryan Houdek Sonicadvance1@gmail.com [0|1|0] http://ln-s.net/-JGa Merge in JP's change to fix audio stuttering with OpenAL.

`git cherry-pick c1fde16` should be used rather than `git apply` because that change the author, date, and commit message

> [09:49] @Sonicadvance1 eh, I just went and changed it. I don't have your remotes in my tree

The commit should be pulled with

```
git pull https://github.com/john-peterson/dolphin-emu audio2
```

The author should be asked if it's accepted to change the author, date or commit message

> [09:51] @Sonicadvance1 Could have done

### Solution

The solution is to skip 0 sample size processing

by changing

```
        if (nSamples < minSamples)
            continue;
```

to

```
        if (nSamples <= minSamples)
            continue;
```

because
- it has an unintended effect on buffering that result in stuttering unless the number of buffers is multiplied by around around 16

The following discussion is oudated because it discuss another change that isn't a problem

### Solution without patch

The solution without the patch is to increase the buffer to 64

Running

SSBM

with

`Dolphin-audio2.exe` in [this folder](https://www.dropbox.com/sh/2z5p162ovqtbn17/y7EfCJd1Oi/file/dolphin/test)

`User/Config/Dolphin.ini` setting

```
DSPHLE = False

DPL2Decoder = False

Latency = 14
```

CPU

```
sudo dmidecode --type processor | grep "Version"
   Version: Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz
```

result in stuttering introduction music

Changing settings to

```
Latency = 30
```

reduce stuttering

to (with patch that increase the maximum buffer size)

```
Latency = 64
```

remove stuttering

### "Generic Software" (wrap_oal.dll) problem

The `User/Config/Dolphin.ini` setting

```
DPL2Decoder = True
```

with device

```
OpenAL Soft
```

has stuttering also when `Latency` is large, f.e.

```
Latency = 32
```

beccause

even when the buffer size is small

```
#define OAL_MAX_BUFFERS         16
```

it writes errors

```
37:20:794 Src\OpenALStream.cpp:257 E[Audio]: Error unqueuing buffers: 0000a002
37:20:795 Src\OpenALStream.cpp:269 W[Audio]: Unable to set 5.1 surround mode.  Updating OpenAL Soft might fix this issue.
37:20:795 Src\OpenALStream.cpp:308 E[Audio]: Error queuing buffers: 0000a004
…[repeated around 50 times]
```

that are associated with subsequent stuttering

### skid_au

> [03:42] @skid_au JPeterson, you there? receiveSamples should be the minimum that can we get away with. usually blocks that the dsp sends or what the timestretcher needs without sounding "hollow"
> 
> [03:58] @skid_au JPeterson, you didn't change the receiveSamples setting in your patch, so it should work the same way, shouldn't it?

The patch change the maximum `soundTouch.receiveSamples` (described in the topic "Solution → `sampleBuffer` size")

> Removing `SIZE_FLOAT` multiplication from this `sampleBuffer` code …
- divide the maximum `soundTouch.receiveSamples` with `SIZE_FLOAT`
- when considering that half of the max is a buffer overflow the maxium without a buffer overflow is divided with `SIZE_FLOAT` / 2

> Changing the code to …

reduce the maxium `soundTouch.receiveSamples` with `OAL_MAX_BUFFERS - numBuffers`

however, this doesn't affect stuttering because changing

```
#define OAL_MAX_BUFFERS         32
unsigned int nSamples = soundTouch.receiveSamples(sampleBuffer, OAL_MAX_SAMPLES * numBuffers);
```

to

```
#define OAL_MAX_BUFFERS         512
unsigned int nSamples = soundTouch.receiveSamples(sampleBuffer, OAL_MAX_SAMPLES * OAL_MAX_BUFFERS);

if (nSamples > OAL_MAX_SAMPLES * numBuffers)
    NOTICE_LOG(COMMON, "%u", nSamples - OAL_MAX_SAMPLES * numBuffers);
```

doesn't alter the stuttering for

```
Latency = 14
```

in SSBM

because
- it print only 13 messages (when the game is started before audio is played), i.o.w. it doesn't have an effect

> [05:26] @skid_au did it stutter before the changes made in your patch?

No there's a significantly increased stuttering with the same Latency setting in the "Clarifying the OpenAL loop" patch

### xbaagx, jordan.c...@gmail.com

[@xbaagx](https://code.google.com/p/dolphin-emu/source/detail?r=adb83cfabe77a04052412d2fd631e341c72319a2)

> This revision broke sound for OpenAl in game "Skate It", now sound is stuttering...

@jordan.c...@gmail.com

> I also experience the stuttering audio from this revision (in Wind Waker). 

For a test
- use the files in the topic "Test"
- or provide a Dropbox link to the files used for the test

because
- using the same files for a test reduce amigbuity regarding build and settings

Inform about

the value of the `User/Config/Dolphin.ini` setting

```
DPL2Decoder =
Latency =
```

the result with the highest latency setting

```
Latency = 30
```

surround sound off

```
DPL2Decoder = False
```

because
- this patch reduce the maximum `soundTouch.receiveSamples` (described in the topic "Solution → `sampleBuffer` size")
- increasing `Latency` increase the maximum `soundTouch.receiveSamples` size

the CPU model by

pasting the output from this command from Cygwin or Linux

```
sudo dmidecode --type processor | grep "Version"
```

providing a link to the CPU model at cpubenchmark.net

because
- the frame rate has an effect on audio
- it's a higher probability that the maximum `soundTouch.receiveSamples` size is used when the frame rate is lower

in which game the audio stutter

where in the game the audio stutter because
- audio stutering isn't noticeable in the Skate It (NTSC) start menu music at 60 FPS with



```
Latency = 30
```

### SizzlingCalamari

> [22:00] <_SizzlingCalamari> i've seen the comment, but increasing the latency only reduces the crackling
> [22:00] <_SizzlingCalamari> it doesn't fully eliminate it
> [22:06] <_SizzlingCalamari> JP, in response, the previous code is better because it does not crackle under the default latency of 14
> [22:06] <_SizzlingCalamari> the new code crackles under all latency settings available
> [22:33] <_SizzlingCalamari> JP, my latency value in the Dolphin.ini config is 14
> [22:33] <_SizzlingCalamari> I can set it to 30 through the ui, but the crackling still remains
> [22:35] <_SizzlingCalamari> JP, yes 30 latency stutters less than 14
> 
> [22:11] <_SizzlingCalamari> JPeterson: i am running windows, but i have an i7 3770K
> 
> [22:25] <_SizzlingCalamari> i'm using an NTSC version of The Legend of Zelda: The Wind Waker
> [22:27] <_SizzlingCalamari> It stutters at the title screen, when the emulator starts up
> 
> [21:53] <_SizzlingCalamari> JPeterson, are you planning on fixing this commit: http://code.google.com/p/dolphin-emu/source/detail?r=adb83cfabe77a04052412d2fd631e341c72319a2
> [21:53] <_SizzlingCalamari> or to anyone, can we revert this

The patch shouldn't be changed based on this analysis because
- changing the patch should be based on a more detailed analysis about the best maximum `soundTouch.receiveSamples` size
- skid_au should comment because he has knowledge about this code

Reply to the questions in the topic "xbaagx, jordan.c...@gmail.com" because
- that information is meaningful to analyse the problem

**Comment by john-peterson** [comment]
## Open device problem

### Problem

Around 1/100 times the device is opened this message is shown

```
PanicAlertT("OpenAL: can't find sound devices");
PanicAlertT("Could not initialize backend %s.", backend.c_str());
PanicAlertT("Sound backend %s is not valid.", backend.c_str());
if(!soundStream) PanicAlert("Error starting up sound stream");
```

### Solution

Use the open instead of non-open OpenAL because
- that can make it easier to locate the problem

**Comment by john-peterson** [comment]
## Adding file monitor file size cache

### Reference
- http://forums.dolphin-emu.org/Thread-1080-avalanche-choppy-audio
- https://code.google.com/p/dolphin-emu/issues/detail?id=6141

### Problem

There are around 5/s buffer underrun messages

```
19:28:589 Src\OpenALStream.cpp:315 W[Audio]: Buffer underrun
19:29:009 Src\OpenALStream.cpp:315 W[Audio]: Buffer underrun
19:29:356 Src\OpenALStream.cpp:315 W[Audio]: Buffer underrun
19:29:767 Src\OpenALStream.cpp:315 W[Audio]: Buffer underrun
19:29:866 Src\OpenALStream.cpp:315 W[Audio]: Buffer underrun
19:30:132 Src\OpenALStream.cpp:315 W[Audio]: Buffer underrun
19:30:243 Src\OpenALStream.cpp:315 W[Audio]: Buffer underrun
19:30:522 Src\OpenALStream.cpp:315 W[Audio]: Buffer underrun
19:30:618 Src\OpenALStream.cpp:315 W[Audio]: Buffer underrun
19:30:889 Src\OpenALStream.cpp:315 W[Audio]: Buffer underrun
```

in 1080 Avalanche

at

`User/Config/Dolphin.ini` setting

```
Latency = 14
Latency = 128
```

because the audio thread is continuously reading from slow media in

```
m_mixer->Mix
```

that takes to long to play continuous audio

### Test

This occur in
- 1080 Avalanche in menu and game (not in FMV), more frequently at the "MATCH RACE" selection than before it

doesn't occur in
- Crazy Taxi

### Solution

Adding a file size cache because

that remove this message from 1080 Avalanche

```
47:41:362 Src\OpenALStream.cpp:315 W[Audio]: Buffer underrun
```

**Comment by john-peterson** [comment]
## Disabling file monitor by default

### Commit

The commit has the same title as this post in https://github.com/john-peterson/dolphin-emu/compare/audio2

### Problem

Described in "Adding file monitor file size cache"

### Solution

Disable the file monitor log `User/Config/Logger.ini`

```
FileMon = False
```

because it cause a significant increase in disk activity for every audio data read

**Comment by john-peterson** [comment]
Only thing I remember of this I eventually made one commit to clean up the audio loop and name many magic numbers. This also fixed a spurious crash that I realized afterwards was the gigantic size of the stack variables. They blatantly overflowed the hardware stack size and Visual Studio had nothing to complain about. And its debugger had nothing sensible to say about the resulting crashes that would occur according to the debugger at random places sometimes near the sound code

I briefly scanned through this issue and it's not something I would propose is added to the library of Congress for safekeeping as culturally significant. I'm looking for things that I clearly want to look at agin that are open issues
